keps/sig-cluster-lifecycle/20190122-Certificates-copy-for-kubeadm-join--control-plane.md

```diff
@@ -0,0 +1,330 @@
+
+The `TokenCleaner` controller will ensure automatic deletion of bootstrap tokens
+and therefore also deletion of the TTL-token.
+
+### Risks and Mitigations
+
+In case the cluster administrator uses the `kubeadm init --upload-certs` flag,
+the `kubeadm-certs` secret will store sensitive cluster certificates. Following
+security measures apply to the `kubeadm-certs` secret:
+
+- the `kubeadm-certs` secret is time lived (with a duration shorter than
+  bootstrap tokens)
+- if an attacker get access to the cluster when the `kubeadm-certs` secret still
+  exists, the cluster certs contained in the secret are protected by RBAC rules
+- if an attacker get access to the cluster when the `kubeadm-certs` secret still
+  exists with an identity that is authorized to read the `kubeadm-certs` secret,
+  the cluster certificates contained in the secret will be protected by the
+  encryption algorithm (without the encryption key, the secret contains a harmless
+  blog of bytes)
+
+The following security measures apply to the encryption key:
+
+- the encryption key will never be stored on the cluster by kubeadm
+- the encryption key will never be transmitted on the wire by kubeadm
+- the encryption key is implicitly time lived (because when the `kubeadm-certs`
+  secret is automatically deleted, the encryption key became useless)
+- it will be possible to hide the encryption key from the `kubeadm init` output
+  by using `--skip-token-print` or similar flag (but this requires the user
+  passing an encryption key to kubeadm using the kubeadm config file), thus
+  avoiding the key improperly registered in logs
+- it will be possible to use a config file for passing the encryption key to
+  `kubeadm join` command, thus avoiding the key improperly registered in
+  shell history
+
+Another possible risk derives from kubeadm allowing the user to fully customize
+control plane components, including also the possibility to remove the TokenCleaner
+from the list of enabled controllers in the kube-scheduler configuration, and as
+a consequence, disable the mechanisms that manages time lived tokens/time lived
+`kubeadm-certs` secret/time lived encryption key.
+
+A possible mitigation action for this risk is to enforce the TokenCleaner in case
+when the user opts-in in the upload-certs workflow defined in this document.
+
+Despite all the above measures, there is still a residual risk in case an attacker
+manages to steal the encryption key from the cluster administrator and to get access
+to the cluster with adequate rights before the `kubeadm-certs` secret is
+automatically deleted.
+
+However, we assume this residual risk could be accepted by cluster administrator
+in favor of the simplified UX offered by the automatic certificates copy during
+the `kubeadm init` / `kubeadm join --control-plane`.
+
+Nevertheless, as additional mitigation action, the potential risk will be clearly
+stated both in the documentation and in the kubeadm output.
```

Comment:
should we still provide a mechanic to cancel the serving of certs in such a scenario?

e.g.
`kubeadm init phases upload-certs cancel`